farm/data_handler/data_silo.py

```diff
@@ -942,3 +942,104 @@ def _split_for_qa(documents, id_index, n_splits=5, shuffle=True, random_state=No
             current_train_set = np.hstack(np.delete(splits, idx, axis=0))
 
             yield current_train_set, current_test_set
+
+
+class DataSiloForNestedCrossVal(DataSiloForCrossVal):
+    """
+    Prepare data silos to do a nested cross validation.
+
+    The outer cross validation splits all data to test and rest.
+    The inner cross validation splits the rest data to train and dev.
+
+    To use this class call ``DataSiloForNestedCrossVal.make()`` which
+    creates a list of DataSiloForCrossVal
+    instances - one for each fold (outer folds times inner folds).
+    """
+
+    def __init__(self, origsilo, trainset, devset, testset):
+        super().__init__(origsilo, trainset, devset, testset)
+
+    @classmethod
+    def make(cls, datasilo, sets=["train", "dev", "test"],
+             n_outer_splits=5, n_inner_splits=5,
+             shuffle=True, random_state=None,
+             stratified=True, n_neg_answers_per_question=1):
+        """
+        Create number of folds data-silo-like objects which can be used
+        to do a nested crossvalidation for training from the
+        original data silo passed on.
+
+        The outer cross validation splits all data to test and rest.
+        The inner cross validation splits the rest data to train and dev.
+        The number of returned data silos is ``n_outer_splits`` times
+        ``n_inner_splits``.
+
+        :param datasilo: the data silo that contains the original data
+        :type datasilo: DataSilo
+        :param sets: which sets to use to create the xval folds (strings)
+        :type sets: list
+        :param n_outer_splits: number of folds for outer cross validation
+        :type n_outer_splits: int
+        :param n_inner_splits: number of folds for inner cross validation
+        :type n_inner_splits: int
+        :param shuffle: shuffle each class' samples before splitting
+        :type shuffle: bool
+        :param random_state: random state for shuffling
+        :type random_state: int
+        :param stratified: if class stratification should be done
+        :type stratified: bool
+        :param n_neg_answers_per_question: number of negative answers per question to include for training
+        :type n_neg_answers_per_question: int
+        """
+
+        if "question_answering" in datasilo.processor.tasks:
+            return cls._make_question_answering(datasilo, sets, n_splits, shuffle, random_state, n_neg_answers_per_question)
```

Comment:
n_splits is not defined and will throw an error. We should nevertheless try to integrate the inner split functionality into DataSiloForCrossVal

Reply:
fixed